packages/firestore/src/core/query.ts

```diff
@@ -165,21 +165,55 @@ export function queryMatchesAllDocuments(query: Query): boolean {
   );
 }
 
-export function getFirstOrderByField(query: Query): FieldPath | null {
-  return query.explicitOrderBy.length > 0
-    ? query.explicitOrderBy[0].field
-    : null;
+// Perform an in depth search on the query to get all inequality filter fields. Returns an empty
+// array if there is no inequality filters.
+export function getInequalityFilterFields(query: Query): FieldPath[] {
```

Comment:
I think we should change the signature of this function to return a `SortedSet<FieldPath>` because
1- callers probably never want duplicates
2- this will automatically sort the fields for them. This may completely remove the need for the lexicographic comparator that you wrote.

i.e.

```js
// Returns the sorted set of inequality filter fields used in this query.
export function getInequalityFilterFields(query: Query): SortedSet<FieldPath> {
  let result = new SortedSet<FieldPath>(FieldPath.comparator);
  query.filters.forEach((filter: Filter) => {
    const inequalityFields = filter
      .getInequalityFilters()
      .map(filter => filter.field);
    inequalityFields.forEach((field: FieldPath) => {
      result = result.add(field);
    });
  });
  return result;
}
```